llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -1517,10 +1518,27 @@ SDValue SelectionDAGLegalize::ExpandInsertToVectorThroughStack(SDValue Op) {
                      BaseVecAlignment);
 }
 
+SDValue SelectionDAGLegalize::ExpandConcatVectors(SDNode *Node) {
+  assert(Node->getOpcode() == ISD::CONCAT_VECTORS && "Unexpected opcode!");
+  SDLoc DL(Node);
+  SmallVector<SDValue, 16> Ops;
+  unsigned NumOperands = Node->getNumOperands();
+  MVT VectorIdxType = TLI.getVectorIdxTy(DAG.getDataLayout());
+  for (unsigned I = 0; I < NumOperands; ++I) {
+    SDValue SubOp = Node->getOperand(I);
+    EVT VectorValueType = SubOp.getValueType();
+    EVT ElementValueType = VectorValueType.getVectorElementType();
+    unsigned NumSubElem = VectorValueType.getVectorNumElements();
+    for (unsigned Idx = 0; Idx < NumSubElem; ++Idx) {
+      Ops.push_back(DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, ElementValueType,
+                                SubOp, DAG.getConstant(Idx, DL, VectorIdxType)));
```

Comment:
A node with `ISD::BUILD_VECTOR` is failing the below assertion.
```
  for (const SDValue &Op : Node->op_values())
    assert((TLI.getTypeAction(*DAG.getContext(), Op.getValueType()) ==
              TargetLowering::TypeLegal ||
            Op.getOpcode() == ISD::TargetConstant ||
            Op.getOpcode() == ISD::Register) &&
            "Unexpected illegal type!");
```
I think this is because these lines will add a node with an illegal value type. Do I also need to make sure that the nodes added here are legal? looking at the code it seems like nodes are legalized in topological order, is there another reason why this is failing?

Reply:
EXTRACT_VECTOR_ELT has the weird property of supporting some implicit casts from the source vector to the result element, you'll probably need to adjust the input.

But yes this is a forever DAG design flaw. GlobalISel does not have this problem

Reply:
Out of curiosity, which unit test produces this assertion failure?

Reply:
`CodeGen/AArch64/sve-streaming-mode-fixed-length-concat.ll` is the one I see failing.